### PR TITLE
Add Vertica connector

### PR DESCRIPTION
**Pull request by ebyhr**
## Description

This connector was originally written for the Starburst Enterprise platform and is now being contributed to Trino.
The documentation will be added by Manfred in a separate PR. 

The code in this PR replaces https://github.com/trinodb/trino/pull/6134

## Release notes

```markdown
## General
* Add Vertica connector. ({issue}`23948`)
```

### CODE REVIEW

File: core/trino-server/src/main/provisio/trino.xml [hunk 1]
```diff
@@ -362,4 +362,10 @@
             <unpack />
         </artifact>
```
Comment (kokosing) [inline core/trino-server/src/main/provisio/trino.xml:363]: I guess docs changes will follow, right?

File: plugin/trino-vertica/src/main/java/io/trino/plugin/vertica/VerticaTableStatisticsReader.java [hunk 2]
```diff
@@ -0,0 +1,488 @@
...
+                            "GROUP BY column_name")
+                    .bind("schema", remoteTableName.getCatalogName().orElse(null))
+                    .bind("table_name", remoteTableName.getTableName())
+                    .map((rs, ctx) -> new AbstractMap.SimpleEntry<>(rs.getString("column_name"), rs.getLong("size")))
+                    .stream()
+                    .collect(Collectors.toMap(AbstractMap.SimpleEntry::getKey, AbstractMap.SimpleEntry::getValue));
+        }
+    }
+
+    public static class Schema
```
Comment (kokosing) [inline plugin/trino-vertica/src/main/java/io/trino/plugin/vertica/VerticaTableStatisticsReader.java:245]: record?

File: plugin/trino-vertica/src/main/java/io/trino/plugin/vertica/VerticaTableStatisticsReader.java [hunk 3]
```diff
@@ -0,0 +1,488 @@
...
+            this.tables = tables;
+        }
+
+        public List<Table> getTables()
+        {
+            return tables;
+        }
+    }
+
+    public static class Table
```
Comment (kokosing) [inline plugin/trino-vertica/src/main/java/io/trino/plugin/vertica/VerticaTableStatisticsReader.java:261]: record?

File: plugin/trino-vertica/src/main/java/io/trino/plugin/vertica/VerticaTableStatisticsReader.java [hunk 4]
```diff
@@ -0,0 +1,488 @@
...
+            return tableName;
+        }
+
+        public List<Column> getColumns()
+        {
+            return columns;
+        }
+    }
+
+    public static class Column
```
Comment (kokosing) [inline plugin/trino-vertica/src/main/java/io/trino/plugin/vertica/VerticaTableStatisticsReader.java:294]: record?

File: plugin/trino-vertica/src/main/java/io/trino/plugin/vertica/VerticaTableStatisticsReader.java [hunk 5]
```diff
@@ -0,0 +1,488 @@
...
+        {
+            return toStringHelper(this)
+                    .add("columnName", columnName)
+                    .add("dataType", dataType)
+                    .add("stats", stats)
+                    .toString();
+        }
+    }
+
+    public static class ColumnStatistics
```
Comment (kokosing) [inline plugin/trino-vertica/src/main/java/io/trino/plugin/vertica/VerticaTableStatisticsReader.java:337]: record?

File: plugin/trino-vertica/src/main/java/io/trino/plugin/vertica/VerticaTableStatisticsReader.java [hunk 6]
```diff
@@ -0,0 +1,488 @@
...
+            return toStringHelper(this)
+                    .add("distinct", distinct)
+                    .add("minValue", minValue)
+                    .add("maxValue", maxValue)
+                    .add("histogram", histogram)
+                    .toString();
+        }
+    }
+
+    public static class Histogram
```
Comment (kokosing) [inline plugin/trino-vertica/src/main/java/io/trino/plugin/vertica/VerticaTableStatisticsReader.java:389]: record?

File: plugin/trino-vertica/src/main/java/io/trino/plugin/vertica/VerticaTableStatisticsReader.java [hunk 7]
```diff
@@ -0,0 +1,488 @@
...
+        public boolean isNan()
+        {
+            if (value == null) {
+                return false;
+            }
+            return value.equals("-nan") || value.equals("nan");
+        }
+    }
+
+    public static class NumericValue
```
Comment (kokosing) [inline plugin/trino-vertica/src/main/java/io/trino/plugin/vertica/VerticaTableStatisticsReader.java:473]: a class for long looks like an overkill

**Review by kokosing — APPROVED** [review]
I checked the code. I feel quite familiar with it so maybe I am not the best person to review it, because I know what I read.

I think this connector is in the good shape to merge.

File: core/trino-server/src/main/provisio/trino.xml (hunk 1 above)
Comment (ebyhr) [reply core/trino-server/src/main/provisio/trino.xml:363]: Right, @mosabua will send docs PR.

File: plugin/trino-vertica/src/main/java/io/trino/plugin/vertica/VerticaTableStatisticsReader.java [hunk 8]
```diff
@@ -0,0 +1,488 @@
...
+            }
+
+            public NumericValue getCount()
+            {
+                return count;
+            }
+        }
+    }
+
+    public static class NullableValue
```
[the remark this answers is not in the document]
Comment (ebyhr) [reply plugin/trino-vertica/src/main/java/io/trino/plugin/vertica/VerticaTableStatisticsReader.java:269]: This class has two constructors. I tried changing to record, but `TestVerticaTableStatistics` started failing. Let me handle in follow-up.

File: plugin/trino-vertica/src/main/java/io/trino/plugin/vertica/VerticaTableStatisticsReader.java (hunk 7 above)
Comment (ebyhr) [reply plugin/trino-vertica/src/main/java/io/trino/plugin/vertica/VerticaTableStatisticsReader.java]: This class is used to retrieve `value` attribute from:
```xml
<distinct value="25"/>
```

```xml
<count value="1"/>
```
It it possible to get the value without this class?

File: core/trino-server/src/main/provisio/trino.xml (hunk 1 above)
Comment (findinpath) [reply core/trino-server/src/main/provisio/trino.xml:363]: https://github.com/trinodb/trino/pull/23967

**Pull request merged**